src/shared/sleep-config.c

```diff
@@ -310,6 +328,19 @@ int find_hibernate_location(HibernateLocation **ret_hibernate_location) {
         if (r < 0)
                 return r;
 
+        /* Users of swapfiles on Btrfs should override swap discovery because
+         * systemd will be unable to correctly calculate offsets.
+         */
+        if (getenv_bool("SYSTEMD_BYPASS_HIBERNATION_MEMORY_CHECK") > 0) {
```

Comment:
This is currently the only place where `SYSTEMD_BYPASS_HIBERNATION_MEMORY_CHECK` is used in the code, and it is used to fail the check, not to pass it, so the current state of the PR appears to be broken (also see https://github.com/systemd/systemd/issues/14249#issuecomment-565739381).

Could you again elucidate your motivation to introduce this environment variable, please? I get that you don't want to arbitrarily skip the memory check, but if users have already configured `resume` and `resume_offset` manually, I think that already demonstrates they know what they are doing. Or to put it differently, in what kind of situation would `resume` and `resume_offset` be configured, but `SYSTEMD_BYPASS_HIBERNATION_MEMORY_CHECK` should be unset?

Reply:
> This is currently the only place where `SYSTEMD_BYPASS_HIBERNATION_MEMORY_CHECK` is used in the code, and it is used to fail the check, not to pass it, 

`SYSTEMD_BYPASS_HIBERNATION_MEMORY_CHECK` is a pre-existing environment variable that was added to allow users to bypass the check for enough memory. It is used to bypass  (to "pass" it, actually) and not to "fail" the check. 

> ...so the current state of the PR appears to be broken (also see [#14249 (comment)](https://github.com/systemd/systemd/issues/14249#issuecomment-565739381)).

See my follow up comment: https://github.com/systemd/systemd/issues/14249#issuecomment-565740678

> Could you again elucidate your motivation to introduce this environment variable, please? 

Again, the environment variable was pre-existing and not introduced here - I'm just using it. It was introduced to allow users to override some of the systemd functionality when they have customized or special swap configurations (which I would argue using a swapfile on Btrfs is) https://github.com/systemd/systemd/commit/490d20e65dd85628f910299a3925bbff466b2e74. 

> I get that you don't want to arbitrarily skip the memory check, but if users have already configured `resume` and `resume_offset` manually, I think that already demonstrates they know what they are doing. Or to put it differently, in what kind of situation would `resume` and `resume_offset` be configured, but `SYSTEMD_BYPASS_HIBERNATION_MEMORY_CHECK` should be unset?

In most situations (ie, using a proper swap partition or using a swapfile on a file system that provides real stat and fiemap information), users do not need to set `SYSTEMD_BYPASS_HIBERNATION_MEMORY_CHECK` - setting it is the exception and using a swapfile on Btrfs seems exceptional - if for no other reason than in order to configure it correctly you need to calculate the offset using non-standard tools.

The situation I think you're talking about specifically is the one where the user has configured `/sys/power/resume` and `/sys/power/resume_offset` (for swap files) but a corresponding entry in `/proc/swaps` could not be identified.

I think you're suggesting that in this situation systemd log a debug message and return successfully to allow hibernation to be attempted. If it fails, the user [hopefully] knows what they are doing so they can fix it.

What I am suggesting is that the user be more explicit that they know what they're doing by setting this environment variable to override some of the protections and discovery that systemd is trying to provide here.

I think what you're suggesting isn't unreasonable  and I think removing the `BYPASS` check in `find_hibernate_location` could be ok. This would eliminate the need to add this environment variable to `systemd-hibernate.service` _in additional_ to `systemd-logind.service` BUT I think you'll still need to set it in `systemd-logind.service` to be very explicit that you wish to override the memory check; this is not about configuring `/sys/power/{resume, resume_offset}` correctly.

Reply:
@diabonas - I just pushed up a change that does what I just described above :point_up:  

You will need to set
```
[Service]
Environment=SYSTEMD_BYPASS_HIBERNATION_MEMORY_CHECK=1
```

In `sytemd-logind.service` (but not also in `systemd-hibernate.service`)

Reply:
> `SYSTEMD_BYPASS_HIBERNATION_MEMORY_CHECK` is a pre-existing environment variable that was added to allow users to bypass the check for enough memory. It is used to bypass (to "pass" it, actually) and not to "fail" the check.

Ah right, I overlooked the existing code in `enough_swap_for_hibernation`, sorry.

> What I am suggesting is that the user be more explicit that they know what they're doing by setting this environment variable to override some of the protections and discovery that systemd is trying to provide here.
> 
> I think what you're suggesting isn't unreasonable and I think removing the `BYPASS` check in `find_hibernate_location` could be ok. This would eliminate the need to add this environment variable to `systemd-hibernate.service` _in addition_ to `systemd-logind.service` BUT I think you'll still need to set it in `systemd-logind.service` to be very explicit that you wish to override the memory check; this is not about configuring `/sys/power/{resume, resume_offset}` correctly.

I agree that the current implementation of this PR is in line with the documentation of the environment variable, so that's fine with me:

> systemd-logind:
> - `$SYSTEMD_BYPASS_HIBERNATION_MEMORY_CHECK=1`  — if set, report that hibernation is available even if the swap devices do not provide enough room for it.

Reply:
Awesome! Glad to hear it worked with your LUKS setup. :+1: